src/transports/ipc.rs

```diff
@@ -0,0 +1,185 @@
+        };
+
+        Response(self.messages_tx.send(message).map(|()| response_rx).map_err(Into::into))
+    }
+}
+
+pub struct Response(Result<oneshot::Receiver<rpc::Value>>);
+
+impl futures::Future for Response {
+    type Output = Result<rpc::Value>;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match &mut self.0 {
+            Err(err) => Poll::Ready(Err(err.clone())),
+            Ok(ref mut rx) => {
+                let value = ready!(futures::Future::poll(Pin::new(rx), cx))?;
+                Poll::Ready(Ok(value))
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct TransportMessage {
+    id: RequestId,
+    request: rpc::Request,
+    response_tx: oneshot::Sender<rpc::Value>,
+}
+
+async fn run_server(unix_stream: UnixStream, messages_rx: mpsc::UnboundedReceiver<TransportMessage>) -> Result<()> {
+    let (socket_reader, mut socket_writer) = unix_stream.into_split();
+    let mut pending_response_txs = BTreeMap::default();
+
+    let mut socket_reader = reader_stream(socket_reader);
+    let mut messages_rx = messages_rx.fuse();
+    let mut read_buffer = vec![];
+
+    loop {
+        tokio::select! {
+            message = messages_rx.next() => match message {
+                Some(message) => {
+                    if pending_response_txs.insert(message.id, message.response_tx).is_some() {
+                        log::warn!("Replacing a pending request with id {:?}", message.id);
+                    }
+
+                    let bytes = helpers::to_string(&message.request).into_bytes();
+                    if let Err(err) = socket_writer.write(&bytes).await {
+                        pending_response_txs.remove(&message.id);
+                        log::error!("IPC write error: {:?}", err);
+                    }
+                },
+                None => {},
+            },
+            bytes = socket_reader.next() => match bytes {
+                Some(Ok(bytes)) => {
+                    read_buffer.extend_from_slice(&bytes);
+
+                    let pos = match read_buffer.iter().rposition(|&b| b == b']' || b == b'}') {
+                        Some(pos) => pos,
+                        None => continue,
+                    };
+
+                    match helpers::to_response_from_slice(&read_buffer[..pos]) {
+                        Ok(response) => {
+                            let remaining_bytes_len = read_buffer.len() - pos;
+                            for i in 0..remaining_bytes_len {
+                                read_buffer[i] = read_buffer[pos + i];
+                            }
+                            read_buffer.truncate(remaining_bytes_len);
+
+                            let outputs = match response {
+                                rpc::Response::Single(output) => vec![output],
+                                rpc::Response::Batch(outputs) => outputs,
+                            };
+
+                            for output in outputs {
```

Comment:
I'd split the content of the loop into a function:
```rust
fn respond(pending_response_tx: &mut _, output: rpc::Output) -> Result<(), ()> {
...
}
```

Which would allow you to use `?` to early exit from the function in case of errors, you could avoid `continue` statements and I think it would make the entire `loop { select!` stuff smaller and easier to follow.
```rust
for output in outputs {
  let _ = respond(&mut pending_response_tx, output);
}
```

Reply:
Done in 7f34638c896677321818bb0c1d5a93a2ad7ab3f4